### PR TITLE
Simplify some Hana code

### PR DESCRIPTION
**Pull request by ldionne**
I'm making this pull request mainly for the sake of showing how some of the logic can be simplified. Note that I wasn't able to compile the tests due to:

```
/Users/ldionne/code/nbdl/src/binders/jsoncpp/Read.hpp:11:9: fatal error: 'jsoncpp/json/json.h' file not found
#include<jsoncpp/json/json.h>
        ^
1 error generated.
```

so I can't assess that it works. Also, you were using `hana::common_t`; you most likely want to use `std::common_type`. `hana::common` is really only useful for implementing the library itself, or when working with potentially heterogeneous types. For example, it allows one to get a common type of `hana::integral_constant_tag<int>` for `hana::int_<1>` and `hana::int_<3>`, whereas (I think) `std::common_type` would normally fail.

I just noticed that my editor trimmed trailing whitespace; sorry about that.

### CODE REVIEW

File: src/Variant.hpp [hunk 1]
```diff
@@ -38,7 +40,7 @@ class Variant
   template<typename T>
   constexpr int typeIdFromType(T type) const
   {
-    return *hana::find(typeIds(), type);
+    return decltype(*hana::find(typeIds(), type))::value;
```
Comment (ricejasonf) [inline src/Variant.hpp:43]: What's the rational for wrapping this in decltype?

**Comment by ricejasonf** [comment]
Thanks! Your advice is always welcome and is certainly worth minor whitespace discrepancies. lol

File: src/Variant.hpp (hunk 1 above)
Comment (ldionne) [inline src/Variant.hpp:43]: Since I use a lambda inside `typeIds`, it can't be `constexpr` anymore.

**Comment by ldionne** [comment]
Hey, you should test this! I wasn't able to compile the tests, so my PR might not even work. Does it work?

**Comment by ricejasonf** [comment]
I just removed the decltype thing and it works. If I remove the *, it says that value is a non-static member function ( .value() )

**Comment by ricejasonf** [comment]
This project is in a state that breaking things is okay. :P

**Comment by ldionne** [comment]
> I just removed the decltype thing and it works. If I remove the *, it says that value is a non-static member function ( .value() )

Right, you're going through the `constexpr` implicit conversion from `hana::integral_constant<int, ...>` to `int`. So even though `*hana::find(typeIds(), type)` is _not_ `constexpr`, it can be converted to a `constexpr int` inside a constant expression. Clever.

**Comment by ricejasonf** [comment]
Incidentally, note that I am using this to convert a run-time index into a type using `overload_linearly`. Variants are useful in general, but this also serves as the bridge between compile-time and run-time. In fact, when I first started to refactor it, I was naively using hana::at with a run-time int. :P

**Comment by ricejasonf** [comment]
It just occurred to me what was going on with non-constexpr `typesIds` being called in a constexpr. Apparently this is valid in C++14.
http://stackoverflow.com/questions/23442942/non-constexpr-calls-in-constexpr-functions

**Pull request merged**